File: Week-Wise-Assignments/Week-2/C/Prog1.c
#include<stdio.h>
#include<stdlib.h>

typedef struct
{
    int *arr;
    int top;
    int capacity;
}Stack;

Stack* createStack(int capacity){
    Stack *stack = (Stack*)malloc(sizeof(Stack));
    stack -> capacity = capacity;
    stack -> top = -1;
    stack -> arr = (int*)malloc(capacity*sizeof(int));
    return stack;
}

void resize(Stack *stack){
    stack->capacity *= 2;
    stack->arr = (int*)realloc(stack->arr, stack->capacity*sizeof(int));
}

void push(Stack *stack, int x){
    if(stack->top + 1 >= stack->capacity)
        resize(stack);
    
    stack->arr[++stack->top] = x;
}

int pop(Stack *stack){
    if(stack->top==-1){
        printf("Queue empty!");
        return -1;
    }

    return stack->arr[stack->top--];
}

int peek(Stack *stack){
    if(stack->top==-1){
        printf("Queue empty!");
        return -1;
    }
    return stack->arr[stack->top];
}

void display(Stack *stack){
    if(stack->top==-1){
        printf("Queue empty!");
        return;
    }

    for(int i = stack->top; i >= 0; i--)
        printf("%d ", stack->arr[i]);

    printf("\n");
}

int main(){
    Stack *stack = createStack(1);
    for(int i = 1;i<6;i++){
        push(stack,i*10);
    }
    display(stack);
    pop(stack);
    display(stack);
    printf("%d\n", peek(stack));
    return 0;
}

File: Week-Wise-Assignments/Week-2/Python/Prog1.py
class Stack:
    def __init__(self):
        self.arr = []
    
    def push(self, x : int):
        self.arr.append(x)
        print(f"{x} appended to Stack")
    
    def pop(self):
        if self.arr:
            return self.arr.pop()
        else:
            print("Empty Stack")
            return None

    def peek(self):
        if self.arr:
            return self.arr[-1]
        else:
            print("Empty Stack")
            return None
    
    def display(self):
        if len(self.arr) != 0:
            print(", ".join([str(x) for x in self.arr]))
        else:
            print("Empty Stack")
    
if __name__ == "__main__": 
    stack = Stack()
    stack.push(20)
    stack.push(30)
    stack.push(40)
    stack.push(50)

    print(stack.pop(), "popped")
    print(stack.pop(), "popped")

    print("Peeking..", stack.peek())
    stack.display()
    stack.pop()
    stack.pop()
    print("Peeking..",stack.peek())
    stack.display()    
